Don't use storage account key for object storage if SP or MI is provided

yvespp opened the pull request:
Implements https://github.com/vmware-tanzu/velero/issues/4267

The plugin no longer uses the storage access key to access the blobs by default. The key is only used if it's configured via `AZURE_STORAGE_ACCOUNT_ACCESS_KEY`. Therefore "Allow storage account key access" can be disabled on the storage account if using MI or SP.

Changes:
* Switched object_store to the newer [azblob](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/storage/azblob#readme)client
* When the storage account access key is not provided delegation SAS instead of service SAS is used.
  * delegation SAS isn't implemented in the azblob client yet so I implemented a simple version in `delegationsas.go`. This code can be delete once it's implemented in [azblob](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/storage/azblob#readme)
  * resourceGroupConfigKey and subscriptionIDConfigKey are no longer needed for the object store
* Added an integration test for the object_storage
  * Only runs when the tag `integration` is specified e.g. `go test ./... -tags=integration -v`
  * The storage account for the test has to be created separatly.
* Updated README
* Updated deps
  * go 1.17
  * go dependencies
  * busybox image

sseago commented on .github/workflows/push.yml:
```diff
@@ -13,10 +13,10 @@ jobs:
     steps:
 
-    - name: Set up Go 1.14
+    - name: Set up Go 1.17
```
velero still uses go 1.16 on the main branch. We should probably do the same here unless there is a specific need for 1.17 for this plugin.

sseago commented on Dockerfile:
```diff
@@ -12,12 +12,12 @@
 # limitations under the License.
 
-FROM golang:1.13-buster AS build
+FROM golang:1.17-buster AS build
```
As above, velero still uses go 1.16 on the main branch. We should probably do the same here unless there is a specific need for 1.17 for this plugin.

sseago reviewed (commented):
Just a question around the golang version for now -- I'll look at this in more detail a bit later.

yvespp commented:
I downgraded it to 1.16 and also added a go setup step to `pr.yaml`, it seems it was using the preinstalled version before.

yvespp commented:
Hi!

Has someone time to review this?
BTW: is there a way to run integration tests against Azure in this repo?

Thanks,
Yves

yvespp commented:
Rebased to master.
Any chance this can be merged?

nxf5025 commented:
This is exactly what we need as well and makes a lot more sense than dealing with access keys when they aren't necessary.

vzabawski commented:
Does the PR also addresses this issue? https://github.com/vmware-tanzu/velero/issues/4930

vzabawski commented:
Awesome! Thank you for your contribution.

chrisob commented:
Reliance on Azure storage account shared keys is preventing my org from using Velero, would be great to see this merged :heart: @ywk253100 @eleanor-millman

anshulahuja98 commented:
HI @yvespp 
I was trying to test out this PR
1 basic flow I was not able to validate - I created a BSL with incorrect params(SA didn't exist)
But the BSL is showing as available. Possibly some underlying layers are not returning errors?
The logs also did not help much debugging

When I created a backup, the following error came in the Backup Status. The BSL was still showing as available

│   completionTimestamp: "2022-09-08T05:12:03Z"                                                                                                                                                                                                                                        ││   expiration: "2022-10-08T05:12:03Z"                                                                                                                                                                                                                                                 ││   failureReason: |                                                                                                                                                                                                                                                                   ││     error checking if backup already exists in object storage: rpc error: code = Unknown desc = ===== RESPONSE ERROR (ErrorCode=AuthorizationPermissionMismatch) =====                                                                                                               ││     Description=, Details: (none)                                                                                                                                                                                                                                                    ││   formatVersion: 1.1.0                                                                                                                                                                                                                                                               ││   phase: Failed                                                                                                                                                                                                                                                                      ││   startTimestamp: "2022-09-08T05:12:03Z"                                                                                                                                                                                                                                             ││   version: 1

ywk253100 commented on velero-plugin-for-microsoft-azure/object_store.go:
```diff
@@ -51,190 +48,207 @@ type containerGetter interface {
-		return "", nil, errors.New("azure subscription ID not found in object store's config or in environment variable")
-	}
+// getCredential from the optional storage account key or via the env
+func getServiceClient(storageAccount, storageKey, storageEndpointSuffix string) (*azblob.ServiceClient, *azblob.SharedKeyCredential, error) {
```
Could you comment when the SharedKeyCredential will be returned and why?

ywk253100 commented:
@yvespp Sorry for the late review.

I have added some comments but haven't reviewed the [delegationsas.go](https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/pull/111/files#diff-1d780eb4040da13fa413bf2491ee1595a92a19977f6aeb27e66cb8a668377042) yet and I will continue to add comments.

mimmi-gjems commented:
Hi! Would be really great if this could be merged sometime soon :) @sseago @ywk253100 @dsu-igeek

vzabawski commented:
I second that. This PR will allow me to start using Velero.

yvespp commented on go.mod:
```diff
@@ -1,70 +1,95 @@
 
 require (
-	github.com/Azure/azure-sdk-for-go v63.4.0+incompatible
+	github.com/Azure/azure-sdk-for-go v66.0.0+incompatible
```
It works for me... I updated dependencies again and could remove the replace section but had to add an exclude because of a ambiguous import.

yvespp commented:
> HI @yvespp I was trying to test out this PR 1 basic flow I was not able to validate - I created a BSL with incorrect params(SA didn't exist) But the BSL is showing as available. Possibly some underlying layers are not returning errors? The logs also did not help much debugging
> 
> When I created a backup, the following error came in the Backup Status. The BSL was still showing as available
> 
> │ completionTimestamp: "2022-09-08T05:12:03Z" ││ expiration: "2022-10-08T05:12:03Z" ││ failureReason: | ││ error checking if backup already exists in object storage: rpc error: code = Unknown desc = ===== RESPONSE ERROR (ErrorCode=AuthorizationPermissionMismatch) ===== ││ Description=, Details: (none) ││ formatVersion: 1.1.0 ││ phase: Failed ││ startTimestamp: "2022-09-08T05:12:03Z" ││ version: 1

Hi @anshulahuja98, do you know how the backup location is verified? I was looking through the code but couldn't find it.

yvespp commented:
@anshulahuja98 the bsl verification should work correctly now. ListCommonPrefixes and ListObjects was returning the wrong error. However, the error is a bit confusing as a 401 (`RESPONSE 401: 401 Server failed to authenticate the request`) is returned instead of a 404.

anshulahuja98 commented:
> @anshulahuja98 the bsl verification should work correctly now. ListCommonPrefixes and ListObjects was returning the wrong error. However, the error is a bit confusing as a 401 (`RESPONSE 401: 401 Server failed to authenticate the request`) is returned instead of a 404.

Great that BSL validation would be fixed.


With regards to error - yes it has become much more confusing now. Earlier it was easier to decode information from the error message - for permissions the client ID and the target principal ID were explicitly mentioned which made permissions fixes straightforward.
I am assuming that with this route - we might not be able to get verbose error messages 

Would be great if we could still find some way to make it verbose.

ywk253100 commented on velero-plugin-for-microsoft-azure/delegationsas.go:
```diff
@@ -0,0 +1,220 @@
+}
+
+// For more information, see https://docs.microsoft.com/en-gb/rest/api/storageservices/create-user-delegation-sas
+type BlobDelegationSASSignatureValues struct {
```
Seems that the upstream has supported the SAS in the latest version https://github.com/Azure/azure-sdk-for-go/tree/sdk/storage/azblob/v0.5.0/sdk/storage/azblob/sas,
could you take a look at it?

yvespp replied:
The code now uses the new version and the custom sas code was removed.

ywk253100 commented:
@yvespp The object store uses the new version of Azure SDK to do the auth and operations in this PR, but the snapshotter still uses the old one. Is it possible to update the snapshotter to use the new SDK as well?

ywk253100 commented on velero-plugin-for-microsoft-azure/object_store.go:
```diff
@@ -51,190 +48,207 @@ type containerGetter interface {
-
-func getStorageAccountKey(config map[string]string) (string, *azure.Environment, error) {
+// get storage account key from env var whose name is in config[storageAccountKeyEnvVarConfigKey].
+func getStorageKey(config map[string]string) (string, error) {
```
How about renaming the function name to `getStorageAccountAccessKey`? It's confused for me what `StorageKey` is?

yvespp replied:
I renamed it to getStorageAccountKey to match the variable names

ywk253100 commented on velero-plugin-for-microsoft-azure/object_store.go:
```diff
@@ -51,190 +48,207 @@ type containerGetter interface {
-	Get(options *storage.GetBlobOptions) (io.ReadCloser, error)
-	Delete(options *storage.DeleteBlobOptions) error
-	GetSASURI(options *storage.BlobSASOptions) (string, error)
+	Get(options *azblob.BlobDownloadOptions) (azblob.BlobDownloadResponse, error)
```
Seems this isn't changed?

yvespp replied:
I changed all but Get, now Get is also changed

yvespp commented:
> @yvespp The object store uses the new version of Azure SDK to do the auth and operations in this PR, but the snapshotter still uses the old one. Is it possible to update the snapshotter to use the new SDK as well?

I moved the snapshotter to the new sdk now and the go-autorest was completely remove. I couldn't test this and it would be good if some else could verify that it works.

ywk253100 commented on go.mod:
```diff
@@ -1,73 +1,87 @@
+// cloud.google.com/go/compute/metadata: ambiguous import: found package cloud.google.com/go/compute/metadata in multiple modules:
+// cloud.google.com/go v0.97.0 (/go/pkg/mod/cloud.google.com/go@v0.97.0/compute/metadata)
+// cloud.google.com/go/compute v1.7.0 (/go/pkg/mod/cloud.google.com/go/compute@v1.7.0/metadata)
+exclude cloud.google.com/go/compute v1.7.0
```
I still cannot develop the code with the IDE Goland, it reports the error
```
GOROOT=/usr/local/go #gosetup
GOPATH=/Users/yinw/go #gosetup
/usr/local/go/bin/go list -modfile=/Users/yinw/go/src/github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/go.mod -m -json -mod=mod all #gosetup
go: k8s.io/kubectl@v0.0.0: invalid version: unknown revision v0.0.0
```
This should be caused by the [kubectl import in azuredisk-csi-driver](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/v1.23.0/go.mod#L108),  adding the follow replace instruction can workaround:
```
replace (
       k8s.io/kubectl => k8s.io/kubectl v0.25.2
)
```

ywk253100 replied:
And for the `exclude cloud.google.com/go/compute v1.7.0` part, I'm suggesting using replace instead:
```
replace (
       cloud.google.com/go => cloud.google.com/go v0.104.0
)
```
Version 0.104.0 enables the go module for the `compute` package.

yvespp replied:
Added like you suggested.

ywk253100 commented:
The BSL validation still doesn't work with the error:
```
RESPONSE 403: 403 This request is not authorized to perform this operation using this permission.
ERROR CODE: AuthorizationPermissionMismatch
```

ywk253100 commented:
And I find another issue: as we change the needed permissions for the customized role, if users use the customized role, this  introduces break change for them

yvespp commented on README.md:
```diff
@@ -175,9 +175,15 @@ There are two ways to specify the role: use the built-in role or create a custom
-   - Storage Account
-      - Microsoft.Storage/storageAccounts/listkeys/action 
-      - Microsoft.Storage/storageAccounts/regeneratekey/action  
+   - Storage Account using storage account access key
```
Removed

ywk253100 commented:
So this is the reason that I got the `AuthorizationPermissionMismatch` error as I still use the old credential with only `Contributor` role.

I'm afraid this breaking change isn't acceptable, this impacts all upgraded users.

We should re-consider the changes and as we're very close to the 1.10 FC date, this PR may be out of the scope of 1.10 if we have no better solution.

yvespp commented:
> So this is the reason that I got the `AuthorizationPermissionMismatch` error as I still use the old credential with only `Contributor` role.
> 
> I'm afraid this breaking change isn't acceptable, this impacts all upgraded users.
> 
> We should re-consider the changes and as we're very close to the 1.10 FC date, this PR may be out of the scope of 1.10 if we have no better solution.

One option would be to fall back to fetching the shared key if accessing the blobs using the default credential fails. But because the plugin doesn't know the container at init time to test access, the client creation has to be delayed until the object store is used. I'll see if I can implement that without making the code too complicated.

I think it would still make sense to clean up the permissions at some point and remove this compatibility code. My question is when it would be allowed to have that breaking change?

ywk253100 commented:
@yvespp Thanks for the quick update.
I think we need a new major version(v2.x) to introduce such kinds of breaking changes, so it may be a long time before we remove the backward compatibility codes.

I'll take a look at the new changes and test them in my local env and run the E2E test, will let you know the result.

ywk253100 commented:
@yvespp I'm a little confused about the Azure roles. The `Contributor` role almost has all the permissions why do we still need the `Storage Blob Data Contributor` role, is there any Azure doc for this part?

yvespp commented:
> @yvespp I'm a little confused about the Azure roles. The `Contributor` role almost has all the permissions why do we still need the `Storage Blob Data Contributor` role, is there any Azure doc for this part?

The clearest doc about this I found is here: [Authorize access to blobs using Azure Active Directory](https://learn.microsoft.com/en-us/azure/storage/blobs/authorize-access-azure-active-directory?source=recommendations#azure-built-in-roles-for-blobs)

> Only roles explicitly defined for data access permit a security principal to access blob data. Built-in roles such as Owner, Contributor, and Storage Account Contributor permit a security principal to manage a storage account, but do not provide access to the blob data within that account via Azure AD. However, if a role includes Microsoft.Storage/storageAccounts/listKeys/action, then a user to whom that role is assigned can access data in the storage account via Shared Key authorization with the account access keys. For more information, see [Choose how to authorize access to blob data in the Azure portal](https://learn.microsoft.com/en-us/azure/storage/blobs/authorize-data-operations-portal).

I think about it this way: `Contributor` is for the control plane and `Storage Blob Data Contributor` for the data plane.

yvespp commented:
I removed  disableStorageAccountKeyFetch again and the storage access key is used only used if subscriptionId and resourceGroup is set.
I think this is easier to configure as users don't need to add and then in a future release remove the disableStorageAccountKeyFetch parameter again.

ywk253100 commented:
@yvespp Thanks for the quick update.

I'm working on some other priorities and have no bandwidth to review the new changes in the next few days. So it cannot meet the FC date(early next week), I'll come back to review it again before RC, but as the changes introduced by this PR are very fundamental it's better to run enough tests for it, so if we cannot do enough review and test before the RC, I'd like to move it to the 1.11 release and make it merged in the early stage of 1.11.

yvespp commented:
Ok for me. 
In case someone following this pull request wants to test, I published a docker image of this here: https://hub.docker.com/r/yvespp/velero-plugin-for-microsoft-azure
A new image will be published if this pull request is updated.

ywk253100 commented on README.md:
```diff
@@ -190,22 +196,26 @@ There are two ways to specify the role: use the built-in role or create a custom
       - Microsoft.Compute/disks/endGetAccess/action
    
-   Use the following commands to create a custom role which has the minimum required permissions:
+   Use the following commands to create a custom role which has the minimum required permissions (when subscriptionId and resourceGroup isn't set):
```
Let's remove `(when subscriptionId and resourceGroup isn't set)` as we recommend users to use the new way and they may don't know what the `subscriptionId` and `resourceGroup`  are if they have no context about this change.

ywk253100 commented on README.md:
```diff
@@ -245,6 +255,17 @@ There are two ways to specify the role: use the built-in role or create a custom
+    az role assignment create --role $AZURE_ROLE --assignee "velero" --scope /subscriptions/$AZURE_SUBSCRIPTION_ID
+    # with built-in roles
+    az role assignment create --role "Contributor" --assignee "velero" --scope /subscriptions/$AZURE_SUBSCRIPTION_ID
+    # not needed when subscriptionId and resourceGroup is set (deprecated)
```
Same comment with above, let's remove `# not needed when subscriptionId and resourceGroup is set (deprecated)`

ywk253100 commented on README.md:
```diff
@@ -284,10 +305,15 @@ Before proceeding, ensure that you have installed and configured [aad-pod-identi
+    az role assignment create --role $AZURE_ROLE --assignee "velero" --scope /subscriptions/$AZURE_SUBSCRIPTION_ID
+    # with predefined roles
+    az role assignment create --role "Contributor" --assignee "velero" --scope /subscriptions/$AZURE_SUBSCRIPTION_ID
+    # not needed when subscriptionId and resourceGroup is set (deprecated)
```
Same comment with above, let's remove `# not needed when subscriptionId and resourceGroup is set (deprecated)`

ywk253100 commented on changelogs/unreleased/000-yvespp:
```diff
@@ -0,0 +1,9 @@
+## All changes
```
Could you refine the changelog and rename the file name with the correct PR number?

ywk253100 commented:
@yvespp I added a few comments about the document, please take a look.
And please also resolve the conflicts and squash the commits

yvespp commented:
> @yvespp I added a few comments about the document, please take a look. And please also resolve the conflicts, rebase and squash the commits

@ywk253100 I addressed you comments, rebased and squashed.

ywk253100 commented:
The changes have passed the end-to-end testing running locally. Thanks @yvespp 

@sseago Could you take a look at this PR when you are available?

jkroepke commented:
Hi all, could the PR released? It looks like this PR is not port of the 1.6.1 version

yvespp commented:
> Hi all, could the PR released? It looks like this PR is not port of the 1.6.1 version.
> 
> Current workaround: Add `Reader and Data Access` Role in addition to `Storage Blob Data Contributor` on the storage account to avoid the error `... does not have authorization to perform action 'Microsoft.Storage/storageAccounts/listKeys/action' over scope ...`
I think the plan is to release this with the next minor version 1.7.0 and Velero 1.11: https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/pull/111#issuecomment-1283665934

If you want to try it out you can use my image which includes this pull request: https://hub.docker.com/r/yvespp/velero-plugin-for-microsoft-azure
It's still on 1.6.0 though.

ywk253100 commented:
@yvespp Thanks
@jkroepke The fix will be available in 1.7. And as the changes introduced by this PR are quite fundamental, we'll back-port the changes to 1.6

vikrantoct7 commented:
If this PR is already merged, may i know the plugin version in which this PR is available?

ywk253100 commented:
@vikrantoct7 @mkemmerz This change isn't available in Velero 1.11, please see https://github.com/vmware-tanzu/velero/issues/5984 for details.

But if you guys take backup without Restic/Kopia, the Velero Azure plugin can be configured as work without listing storage account access keys.
